docs/ref/optimizers/ceres_tutorial.md

```diff
@@ -0,0 +1,205 @@
+
+
+## How do I define an analytical cost function?
+The implementation for an analytical cost function is more involving compared to a numerical or automatic differentiated cost function. First we have to implement the analytical cost function using `ceres::SizedCostFunction`:
+
+    class QuadraticCostFunction
+      : public SizedCostFunction<1 /* number of residuals */,
+                                 1 /* size of first parameter */> {
+    public:
+      virtual ~QuadraticCostFunction() {}
+      virtual bool Evaluate(double const * const * parameters,
+                            double *residuals,
+                            double **jacobians) const {
+        double x = parameters[0][0];
+        // f(x) = 10 - x.
+        residuals[0] = 10 - x;
+
+        // f'(x) = -1. Since there's only 1 parameter and that parameter
+        // has 1 dimension, there is only 1 element to fill in the
+        // jacobians.
+        //
+        // Since the Evaluate function can be called with the jacobians
+        // pointer equal to NULL, the Evaluate function must check to see
+        // if jacobians need to be computed.
+        //
+        // For this simple problem it is overkill to check if jacobians[0]
+        // is NULL, but in general when writing more complex
+        // CostFunctions, it is possible that Ceres may only demand the
+        // derivatives w.r.t. a subset of the parameter blocks.
+        if (jacobians != NULL && jacobians[0] != NULL) {
+          jacobians[0][0] = -1;
+        }
+        return true;
+      }
+    };
+
+Keypoints:
+
+- Instead of overriding the `()` operator, with a `SizedCostFunction` we override the `Evaluate()` method.
+- `parameters` contains all the data for the cost function.
+- The jacobian for the cost function should be calculated when `jacobians` is not `NULL`
+- Dimensions for residuals and parameters are now defined in the inherited class typecast
+
+After defining the cost function to use it you simply do:
+
+    ceres::CostFunction *cost_function = new AnalyticalCostFunction;
+
+
+## How do I load the optimization data?
+The examples we have seen until now are simple optimization problems with no data. The original purpose of least squares and non-linear least squares analysis was fitting curves to data. It is only appropriate that we now consider an example of such a problem. Lets consider a new cost function:
+
+\begin{equation}
+    y = e^{mx + c}
+\end{equation}
+
+We can implement the cost functor for the above cost function as:
+
+    struct ExponentialResidual {
+      ExponentialResidual(double x, double y)
+          : x_(x), y_(y) {}
+
+      template <typename T>
+      bool operator()(const T* const m, const T* const c, T* residual) const {
+        residual[0] = T(y_) - exp(m[0] * T(x_) + c[0]);
+        return true;
+      }
+
+    private:
+      // Observations for a sample.
+      const double x_;
+      const double y_;
+    };
+
```

Comment:
Are we using this->x or x_, y_ for members? I know this is just docs, so might not need to match.